src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -0,0 +1,134 @@
+            operationHasOperator(operation, SQL_COUNT_OPERATOR); 
+            operationHasOperator(operation, SQL_AVERAGE_OPERATOR);
+            operationHasOperator(operation, SQL_MAX_OPERATOR)
+      ).
+
+.decl isSqlValue(value: AccessPath)
+isSqlValue(value) :- isSqlOperation(operation),
+      operationHasResult(operation, value).
+
+// Keeping track of all initial inputs. 
+.decl isInitialInput (ingress_point: AccessPath)
+isInitialInput(as(ingress_point, AccessPath)) :- isOperation(operation), 
+      operationHasOperator(operation, CORE_INPUT_OPERATOR), 
+      operationHasResult(operation, ingress_point), 
+      operationHasAttribute(operation, [_, $StringAttributePayload(_)]).
+
+// Helper functions to flatten input/operand lists. 
+.decl intermediateFlatten(list:OperandList, sublist:OperandList, list_element:AccessPath)
+
+intermediateFlatten(list, sublist, list_element) :- isOperation(operation), 
+      operationHasOperandList(operation, list), 
+      list = [list_element, sublist].
+intermediateFlatten(list, sublist, list_element) :- intermediateFlatten(list, [list_element, sublist], _).
+
+.decl flatten(list:OperandList, list_element:AccessPath)
+flatten(list, list_element) :- intermediateFlatten(list, _, list_element).
+
+// Flattening each individual input per value. 
+.decl isSqlInput(operation: Operation, sql_input: AccessPath)
+
+isSqlInput(operation, sql_input) :- isSqlOperation(operation), 
+      operationHasOperandList(operation, input_list), 
+      flatten(input_list, sql_input). 
+
+// Keeping track of sensitivity score per each input at index. 
+// This is used to calculate the maximum input sensitivity. 
+// It's decomposed into base and inductive cases because recursive 
+// aggregates are not supported yet for the compiler case in souffle
+.decl sqlOperationHasSensitivityScoreAtIndex(operation: Operation, sensitivity_score: number, index: number) 
+
+// Base cases (index 0).
+// Sesnitivity score for initial inputs is 1. 
+sqlOperationHasSensitivityScoreAtIndex(operation, 1, 0) :- isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, 0),
+      isInitialInput(sql_input).
+
+// For all other inputs, we get the final sensitivity score. 
+sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, 0) :- isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, 0),
+      sqlValueSensitivity(sql_input, sensitivity_score).      
+
+// Inductive cases.  
+// We keep track at the score at each index, and only update if the new input score is higher. 
+sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index) :- 
+      isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, index),
+      isSqlInput(operation, sql_input), 
+      ( 
+            sqlValueSensitivity(sql_input, input_sensitivity_score), 
+            sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index-1), 
```

Comment:
In general, try to express your inductive cases using `+1` rather than `-1`. So that would make this rule look more like:

```
sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, prev_index + 1) :-
...
  (
    sqlValueSensitivity(sql_input, input_sensitivity_score),
    sqlOperationHasSensitivityScoreAndIndex(operation, sensitivyt_score, prev_index)
```

It's just more in line with the classical inductive structure of "Prove for 0, then prove that true for N -> true for N + 1".

Reply:
Done.